### ISSUE TITLE
bug(query): listener removal does not include query parameters

### ISSUE THREAD

**Issue by minusoneman** [prescottprue/react-redux-firebase#294]
**Do you want to request a *feature* or report a *bug*?**  

Potentially a race-condition, but there might be a another explanation that I don't see, or perhaps it's a more general Component Lifecycle/redux issue.

Seems similar to #234, so I wonder if this could be still be a problem in 2.0.0-beta-10?

**What is the current behavior?**

1. Mount a firebaseConnect enabled container (Container1) with a given path and custom storeAs value
2. Navigate to a new firebaseConnected container (Container 2) using the same path
3. Container1 unmounts, but there seems to be a delay before UNSET_LISTENER is dispatched
4. Container2 mounts and SET_LISTENER is dispatched with a path identical to Container1
5. UNSET_LISTENER is dispatched from Container1, conflicting with the SET_LISTENER just dispatched from Container2

As a result, the listener set in Container1 is not unWatched, which can build up over time and hog a lot of resources.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem via [codesandbox](https://codesandbox.io/) or similar.**

```javascript
// Container1
export default compose(
  connect(state => ({ organisationKey: getOrgKey(state) }),
  firebaseConnect(({ organisationKey }) => ([
    { 
        type: 'value', //This is not a problem when called with a 'once' event
        storeAs: 'members',
        path: `/organisations/${organisationKey}/members`
     }
   ])),
  connect(mapStateToProps),
)(Container1)

// Container2
export default compose(
  connect(state => ({ organisationKey: getOrgKey(state) }),
  firebaseConnect(({ organisationKey }) => ([
    { 
        storeAs: 'differentStoreAs', //Doesn't seem to matter if I use the same storeAs or not
        path: `/organisations/${organisationKey}/members` //Same path as Container1
     }
   ])),
  connect(mapStateToProps),
)(Container2)
```

Initially I thought it was a issue with lib/utils/query.js

```javascript
export const unsetWatcher = (firebase, dispatch, event, path, queryId = undefined) => {
  let id = queryId || getQueryIdFromPath(path, event) || getWatchPath(event, path)
  path = path.split('#')[0] <- How can firebase turn off a listener with a path that includes a @storeAs key?
```
so I tried unWatching using just the base path (without storeAs key)
```javascript
export const unsetWatcher = (firebase, dispatch, event, path, queryId = undefined) => {
  let id = queryId || getQueryIdFromPath(path, event) || getWatchPath(event, path)
  path = path.split('#')[0]
  path = path.split('@')[0] <- Attempting the turn off the listener using a base path
```

Which highlighted the problem, as the firebase SDK attempts to turn off the listener for a non-existing path.

**What is the expected behavior?**

Container1 should dispatch UNSET_LISTENERS before Container2 dispatches SET_LISTENERS. Maybe it's a problem with how I deal with Navigation, but I figured I'd bring up this issue for discussion.

**Which versions of dependencies, and which browser and OS are affected by this issue? Did this work in previous versions or setups?**
```
react-redux-firebase v2.0.0-beta-10
react-native-firebase (tested on Android) v3.0.3
react-redux v5.0.5
react-navigation v1.0.0-beta-13
redux-saga v0.15.6
```

**Comment by prescottprue** [comment]
@minusoneman Interesting. Glad you brought it up with details and code snippets. Going to try to reproduce myself, but you may be correct with the routing being the cause.

Open to others ideas on this front if anyone else is experiencing it.

**Comment by minusoneman** [comment]
Thanks @prescottprue !

I did some more testing, and it indeed seems like the Component2 (using Container2) is mounting before Component1 is unmounted, which explains the ordering of the dispatches.

Should react-redux-firebase v2 (currently) be able to handle unwatching uniquely assigned on()'s?

I'll dig deeper into the code, but this bit (utils/query.js:104) seems odd to me:

```javascript
if (event !== 'first_child' && event !== 'once') {                                                           
        firebase.database().ref().child(path).off(event)                                                           
        // TODO: Remove config.distpatchOnUnsetListener                                                            
}
```

Is this where the actual unWatching call to firebase is happening? If we're only relying on path and event type, it could explain the weird behaviour I'm seeing (as multiple listeners are attached to the same path).

from the firebase docs:

> Detach a callback previously attached with on(). Note that if on() was called multiple times with the same eventType and callback, the callback will be called multiple times for each event, and off() must be called multiple times to remove the callback. Calling off() on a parent listener will not automatically remove listeners registered on child nodes, off() must also be called on any child listeners to remove the callback.

If I'm reading this correctly, and we currently are not differentiating between multiple on()'s on the same path, I'd be happy to try take a stab a PR that could make slightly more robust.

Appreciate your thoughts on this.

**Comment by prescottprue** [comment]
@minusoneman That is the line where the listener is being removed.

What you are saying with the listener being detached based on just the path and not the query params seems exactly right. Hopefully the change is as easy as passing any query params along as we are making the off call.

Totally open to a pull request if you can get to it. I have been really busy with the firestore integration stuff, so it is great to get all the help I can :1st_place_medal: .

**Comment by minusoneman** [comment]
@prescottprue happy to help! Not sure it's as easy as simply providing the query params to the off() call:

Let me try and outline a (possible?) solution, and let's see if it makes sense to you.

As per the firebase documentation (https://firebase.google.com/docs/reference/js/firebase.database.Reference), in order to cleanly turn off a listener, we need to keep track of the actual callback (snapshot) used when calling on(). Right now that is defined (but not tracked) in actions/query.js:117

```javascript
query.on(type, (snapshot) => {                                                                                 
  let data = (type === 'child_removed') ? undefined : snapshot.val()                                           
   const resultPath = storeAs || (type === 'value') ? path : `${path}/${snapshot.key}` 
  //...
```

So there is no easy way (that I can see), to reference this callback when we want to turn off the listener.

```javascript
// Ideal situation. callback is passed on to unWatchEvent
ref.off(event, callback)
```

There is an 'older' example in the react-native-firebase on how to deal with this https://github.com/invertase/react-native-firebase/blob/master/docs/v2/modules/database.md

where they suggest the callback is defined in the component. Obviously, this is quite different from the current implementation.

So how can we store the (reference) to the callbacks? One naive solution I might try is to move the callback to the component level (yuk), and pass it to the unWatchEvents method (called in componentWillUnmount), but that's probably not a trivial task.

Any other suggestions on how we can keep track of the callback (saving is in redux?) would be much appreciated.

**Comment by prescottprue** [comment]
So storing the callbacks for unregister is actually what I am currently doing for the Firestore stuff. What I have been doing is just store them on the internal Firebase instance that is passed around (on the `_` parameter).

I'll try some stuff out today/tomorrow and get back to you

**Comment by prescottprue** [comment]
@minusoneman Thanks for the update, been thinking a lot about this.

I don't of a method or listing listeners/queries, but I usually turn on `enableLogging: true` so you can see the debugging info that comes from the firebase lib.

Part of why I have been intending to store info about the watchers/listeners in state is so that we could easily see a list of them like you are mentioning. There has also been requests for an option to create a listener using `firebaseConnect` that will remain outside of unmounts, and the only good way to do that would be if we are tracking them somewhere (or else listeners could easily get out of hand).

Going to try out removing the listeners with a stored subscriber like we are mentioning (how I have been doing it with firestore stuff too) and see if I experience them not detaching as well.

**Comment by minusoneman** [comment]
@prescottprue I agree that keeping a list of active listeners in redux would be great. If only the firebase SDK allowed for listeners to be turned on/off with an self-assigned identifier (instead of a callback), that would be great.

I managed to get some debug information out (since react-native-firebase, there is no easy way to initialize a firebase instance in JS), so I have currently hardcoded logging for events on/off (in Android).

Odd thing, with my modified firebaseConnect.js, I can see that a call is made to turn off the listeners (in Android), but even then, I still see updates to @@reactReduxFirebase/SET to paths that should have been correctly turned off after unmounting. Is there any 'internal' listener in react-redux-firebase, that might be lingering (firebase.listeners is empty), or is the react-redux-firebase simply reacting to changes set (incorrectly) from the firebase instance?

**Comment by prescottprue** [comment]
@minusoneman I am talking about storing the callback itself (the function) on the internal instance that is passed around (which for lack of a better term I called "firebase"). It is an extended version of Firebase that has a `_` parameter which is used to store internals.

As for still seeing the data coming back it is hard to tell without looking deeper. `react-redux-firebase` will only react to listeners set through it (i.e. `watchEvent` or through `firebaseConnect` which uses `watchEvent`).

**Comment by minusoneman** [comment]
@prescottprue I'm trying to wrap my head around this. If we're going to store the callback in the 'firebase' instance (and not the component), how will we know when to turn off the listener? 

I like that the current implementation clean up at componentWillUnmount(),  but that will still require some sort of mapping from component(s) <=> callback(s) (no matter how or where we store it), so we can turn off the correct listeners. This is why I initially thought we 'simply' store the callbacks directly in the component instance.

Thoughts?

**Comment by prescottprue** [comment]
We will know how to turn it off by its key that it is stored on the internal Firebase object (which can be the queryId).

`componentWillUnmount` will still be able to call `unWatchEvents` with the list of query config objects (stored on the component) to detach listeners, but the listeners themselves (the functions) will be stored on the internal object.

That way, in the future we can support combining of listeners or optionally leaving them attached on unmount. This is actually what I am doing for the Firestore listeners.

**Comment by boaticus** [comment]
Just to add another vote/voice: I'm seeing this same exact issue in my project.

Specifically, if I browse to two different routes back-to-back in my project, with both routes having at least one components that are watching the same Firebase path, the UNSET_LISTENER for the previous component is fired _after_ the SET_LISTENER for the new component.

I'm on 2.0.0-beta.14.

Using or not using storeAs doesn't seem to have any effect on the issue.

@minusoneman's steps to reproduce code snippet adequately illustrates my use case, as well.

**Comment by prescottprue** [comment]
@boaticus Great to hear. This is the next highest bug on the list, so hoping to have an update on it soon.

**Comment by minusoneman** [comment]
Just a note to those who might experience this using react-native-firebase. A new version was just pushed, that (in my initial testing) seems partly solve the issue ( see https://github.com/invertase/react-native-firebase/issues/532)

**Comment by prescottprue** [comment]
Have a fix on the `v2.0.0-beta.15` branch I am working on (not yet published to npm) that you can feel free to test out by npm linking. I was able to reproduce the unset/reset as described, then was able to prevent it by including `storeAs` in the `queryId`:

**Before**
![react-redux-firebase_material-ui_demo](https://user-images.githubusercontent.com/2992224/32271337-ab5601f8-beb6-11e7-81d4-0bad626ea2f0.png)

**After**
![react-redux-firebase working](https://user-images.githubusercontent.com/2992224/32271341-adc0449e-beb6-11e7-9213-83486964b635.png)

I also confirmed through enabling Firebase logging that the listeners were being unset as expected when leaving.

#### The Catch
Currently listeners are being checked for removal by their `queryId` **but** they are [being removed based on their path](https://github.com/prescottprue/react-redux-firebase/blob/v2.0.0/src/utils/query.js#L105). That was done because of how Firebase automatically removes all listeners of at that path of that event type ([described in Firebase's docs for `off`](https://firebase.google.com/docs/reference/node/firebase.database.Reference#off) if a callback is not passed.

I *believe I have found a solution* for this as well, but it would mean a bit of a re-write as far as listeners are concerned. Open to doing this and squeezing it into v2, but not at all sure on a timeline, so I will get back to you on that.

**Comment by prescottprue** [comment]
The initial solution was included in [the v2.0.0-beta.15](https://github.com/prescottprue/react-redux-firebase/releases/tag/v2.0.0-beta.15). Still continuing to investigate the remove listener by callback instead of id route (hoping to have it ready/tested for an upcoming v2.0.0 pre-release).

Please let me know if this does not fix things for you or if it causes any issues. Issues with listener management are always very high on the priority list, thank you for reporting.

**Comment by minusoneman** [comment]
Thanks a lot @prescottprue. Have been testing for a couple of days, and it seems to work as expected.
